2nd try at using a vanity GCR name

thockin opened the pull request:
The 2nd commit here is the changes relative to the reverted PR.  Please focus review attention on that.

This is the 2nd attempt.  The previous try (#57573) was reverted while we
figured out the regional mirrors (oops).
    
New plan: k8s.gcr.io is a read-only facade that auto-detects your source
region (us, eu, or asia for now) and pulls from the closest.  To publish
an image, push k8s-staging.gcr.io and it will be synced to the regionals
automatically (similar to today).  For now the staging is an alias to
gcr.io/google_containers (the legacy URL).
    
When we move off of google-owned projects (working on it), then we just
do a one-time sync, and change the google-internal config, and nobody
outside should notice.
    
We can, in parallel, change the auto-sync into a manual sync - send a PR
to "promote" something from staging, and a bot activates it.  Nice and
visible, easy to keep track of.

xref https://github.com/kubernetes/release/issues/281

TL;DR:
  *  The new `staging-k8s.gcr.io` is where we push images.  It is literally an alias to `gcr.io/google_containers` (the existing repo) and is hosted in the US.
  * The contents of `staging-k8s.gcr.io` are automatically synced to `{asia,eu,us)-k8s.gcr.io`.
  * The new `k8s.gcr.io` will be a read-only alias to whichever regional repo is closest to you.
  * In the future, images will be promoted from `staging` to regional "prod" more explicitly and auditably.

 ```release-note
Use "k8s.gcr.io" for pulling container images rather than "gcr.io/google_containers".  Images are already synced, so this should not impact anyone materially.
    
Documentation and tools should all convert to the new name. Users should take note of this in case they see this new name in the system.
```

cblecker commented on CHANGELOG-1.4.md:
```diff
@@ -177,15 +177,15 @@ filename | sha256 hash
 
 * kube-apiserver now drops unneeded path information if an older version of Windows kubectl sends it. ([#44586](https://github.com/kubernetes/kubernetes/pull/44586), [@mml](https://github.com/mml))
-* Bump gcr.io/google_containers/glbc from 0.8.0 to 0.9.2. Release notes: [0.9.0](https://github.com/kubernetes/ingress/releases/tag/0.9.0), [0.9.1](https://github.com/kubernetes/ingress/releases/tag/0.9.1), [0.9.2](https://github.com/kubernetes/ingress/releases/tag/0.9.2) ([#43098](https://github.com/kubernetes/kubernetes/pull/43098), [@timstclair](https://github.com/timstclair))
+* Bump k8s.gcr.io/glbc from 0.8.0 to 0.9.2. Release notes: [0.9.0](https://github.com/kubernetes/ingress/releases/tag/0.9.0), [0.9.1](https://github.com/kubernetes/ingress/releases/tag/0.9.1), [0.9.2](https://github.com/kubernetes/ingress/releases/tag/0.9.2) ([#43098](https://github.com/kubernetes/kubernetes/pull/43098), [@timstclair](https://github.com/timstclair))
```
Is it worth updating historial changelogs?

thockin replied:
I waffle on this.  The image may eventually cease to exist at the old URL?  Or maybe we simply say it won't?

cblecker replied:
My thought is, the application configuration and manifests for, say 1.4, will still be looking for the image at `gcr.io/google_containers/glbc`. Modifying the changelog, doesn't go back and create a new patch release of 1.4 updating the config/manifests to look at the new place.

I'd say it would be worth doing for any version we are actually going to go back and change. If this is just for 1.10 forward, then I'd suggest leaving the historical change logs be.

thockin replied:
ACK

cblecker commented:
FYI cross is failing across the board: #57822

dekkagaijin commented:
Pushes should now be made to `staging-k8s.gcr.io`. `k8s.gcr.io` will be turned into a read-only alias for the local replicas (i.e. `(asia|eu|us)-k8s.gcr.io`).

thockin committed:
Switch to k8s.gcr.io vanity domain

This is the 2nd attempt.  The previous was reverted while we figured out
the regional mirrors (oops).

New plan: k8s.gcr.io is a read-only facade that auto-detects your source
region (us, eu, or asia for now) and pulls from the closest.  To publish
an image, push k8s-staging.gcr.io and it will be synced to the regionals
automatically (similar to today).  For now the staging is an alias to
gcr.io/google_containers (the legacy URL).

When we move off of google-owned projects (working on it), then we just
do a one-time sync, and change the google-internal config, and nobody
outside should notice.

We can, in parallel, change the auto-sync into a manual sync - send a PR
to "promote" something from staging, and a bot activates it.  Nice and
visible, easy to keep track of.

thockin commented on cluster/common.sh:
```diff
@@ -478,7 +478,7 @@ function stage-images() {
 
-  if [[ "${KUBE_DOCKER_REGISTRY}" == "gcr.io/"* ]]; then
+  if [[ "${KUBE_DOCKER_REGISTRY}" =~ "gcr.io/" ]]; then
     local docker_push_cmd=("gcloud" "docker")
```
Once staging becomes a default, we only need the -- part, right?

thockin commented:
/retest

thockin commented:
Un-holding this.  I think it works now.  Let's see what tests say.  PTAL

thockin commented:
/retest

cblecker commented on api/openapi-spec/swagger.json:
```diff
@@ -75018,7 +75018,7 @@
     "properties": {
      "names": {
-      "description": "Names by which this image is known. e.g. [\"gcr.io/google_containers/hyperkube:v1.0.7\", \"dockerhub.io/google_containers/hyperkube:v1.0.7\"]",
+      "description": "Names by which this image is known. e.g. [\"k8s.gcr.io/hyperkube:v1.0.7\", \"dockerhub.io/google_containers/hyperkube:v1.0.7\"]",
```
Does the dockerhub.io/google_containers make sense here? (even if it's still just an example)

thockin replied:
it never really made sense, IMO

cblecker reviewed (approved):
This looks good to me. Can't find any other meaningful references to google[-_]containers in k/k. Tests are all passing, including cross build.

cblecker commented:
/lgtm
/hold

This looks good. Holding to ensure this doesn't merge before @thockin is ready. Feel free to cancel the hold whenever.

cblecker commented:
/lgtm

@thockin Are you good with this merging whenever tests pass?

thockin commented:
@cblecker yes

cblecker commented:
/hold cancel
This should be good then once https://github.com/kubernetes/kubernetes/issues/59223 is fixed.

thockin commented:
/retest

cblecker commented on cluster/images/kubemark/Makefile:
```diff
@@ -15,8 +15,8 @@
 
-REGISTRY?=gcr.io
-PROJECT?=google_containers
+REGISTRY ?= gcr.io
```
Did this get lost in a rebase? Should this be staging-k8s.gcr.io?

thockin replied:
ACK, pushed!  Good catch.

cblecker commented:
/lgtm
Over four hundred lines of the same change starts to blur together after awhile hah. Hopefully this works 🤞. Should be a good time to get it merged too, as the SQ is pretty empty.

cblecker commented on cluster/images/kubemark/Makefile:
```diff
@@ -15,18 +15,17 @@
 
-REGISTRY?=gcr.io
-PROJECT?=google_containers
+REGISTRY ?= staging-k8s.gcr.io
```
Looks like something is overriding this in testing:
```
I0206 21:23:43.775] gcloud docker -- push gcr.io/kubemark
I0206 21:23:44.496] The push refers to a repository [gcr.io/kubemark]
W0206 21:24:34.697] unknown: Pushing to root-level images is disabled. See https://cloud.google.com/container-registry/docs/support/deprecation-notices
W0206 21:24:34.760] make: *** [gcloudpush] Error 1
```

https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/57824/pull-kubernetes-kubemark-e2e-gce/75252/?log#log

cblecker replied:
Maybe it's this: https://github.com/kubernetes/kubernetes/blob/ebffd76acf84d14d7d72fda747e98429d0932895/cluster/images/kubemark/BUILD#L15-L16

dekkagaijin replied:
Doesn't look like `docker_build` messes with env vars:
https://github.com/bazelbuild/rules_docker/blob/master/docker/docker.bzl
https://github.com/bazelbuild/rules_docker/blob/master/container/push.bzl

BenTheElder replied:
adding here: it's not this, that job is using the makefile build. I've poked through test-infra a bit and I don't see this env being set for this job though...

thockin replied:
I know how to fix the makefile part here, I don't know how to fix the bazel part.  Pushing now.  PTAL at bazel equiv.

thockin replied:
pushed another, trying to get bazel.

Bazel doesn't seem to have the same semantic as make here.  Make allows "if this variable is not set, use a default".  So I just repeated the default value.  Sigh.  Someone who knows this script can maybe take a look later ( @shyamjvs )

thockin commented:
/retest

stp-ip commented:
/test pull-kubernetes-kubemark-e2e-gce

cblecker commented:
/lgtm cancel

dekkagaijin commented on test/kubemark/start-kubemark.sh:
```diff
@@ -221,7 +221,7 @@ function create-and-upload-hollow-node-image {
   RETRIES=3
   for attempt in $(seq 1 ${RETRIES}); do
-    if ! bazel run //cluster/images/kubemark:push --define PROJECT="${PROJECT}"; then
+    if ! bazel run //cluster/images/kubemark:push --define REGISTRY="${FULL_REGISTRY}"; then
```
This works with the current implementation of docker_build, since it's just concatenating the strings under the covers, but it would be good to be able to be able to assign `$PROJECT/` as a base repository for `kubemark` in the BUILD target, in case that ever changes in the future.

thockin replied:
?  PROJECT is a faux construct, AFAICT.  gcr.io/google_containers is the registry.  If you want something else, specify the whole thing.

dekkagaijin replied:
In the Docker sense, 'registry' == `hostname[:port]` and 'repository'  == `path` of the URLesque image name: https://docs.docker.com/registry/spec/api/#overview
I'm worried that the a future implementation of `docker_push` might make use of that assumption.

Suggestions added in the relevant places...

cblecker commented on cluster/images/kubemark/BUILD:
```diff
@@ -12,8 +12,8 @@ docker_build(
-    registry = "gcr.io",
-    repository = "$(PROJECT)/kubemark",
+    registry = "$(REGISTRY)",
+    repository = "kubemark",
```
@ixdy @BenTheElder Does this look okay from the bazel side?

BenTheElder replied:
This looks like valid bazel, seems fine to me.

thockin commented:
/retest

BenTheElder commented:
/retest
this looks like a known flake, @mikedanese did your timeout bump go in?

cblecker commented:
@thockin If the test failure is a flake, you should be good to squash

BenTheElder commented:
All green ✅

dekkagaijin commented on cluster/images/kubemark/BUILD:
```diff
@@ -12,8 +12,8 @@ docker_build(
-    registry = "gcr.io",
-    repository = "$(PROJECT)/kubemark",
+    registry = "$(REGISTRY)",
+    repository = "kubemark",
```
```
registry = "$(REGISTRY)",
repository = "$(REPO_PATH_BASE)kubemark",
```

dekkagaijin commented on test/kubemark/start-kubemark.sh:
```diff
@@ -193,14 +193,14 @@ function create-and-upload-hollow-node-image {
   RETRIES=3
   for attempt in $(seq 1 ${RETRIES}); do
-    if ! REGISTRY="${CONTAINER_REGISTRY}" PROJECT="${PROJECT}" make "${KUBEMARK_IMAGE_MAKE_TARGET}"; then
+    if ! REGISTRY="${FULL_REGISTRY}" make "${KUBEMARK_IMAGE_MAKE_TARGET}"; then
```
`if ! REGISTRY="${REGISTRY}" REPO_PATH_BASE="${REPO_PATH_BASE}" make "${KUBEMARK_IMAGE_MAKE_TARGET}"; then`

dekkagaijin commented on test/kubemark/start-kubemark.sh:
```diff
@@ -457,6 +457,15 @@ write-local-kubeconfig
+else
+  FULL_REGISTRY=staging-k8s.gcr.io
+fi
+
```
```
# TODO: Simplify. The caller should have to pass a full registry or nothing.
if [[ -n "${CONTAINER_REGISTRY}" ]]; then
  REGISTRY="${CONTAINER_REGISTRY}"
else
  REGISTRY=staging-k8s.gcr.io
fi
if [[ -n "${PROJECT}" ]]; then
 REPO_PATH_BASE="${PROJECT}/"
fi
```

dekkagaijin commented on test/kubemark/start-kubemark.sh:
```diff
@@ -221,7 +221,7 @@ function create-and-upload-hollow-node-image {
   RETRIES=3
   for attempt in $(seq 1 ${RETRIES}); do
-    if ! bazel run //cluster/images/kubemark:push --define PROJECT="${PROJECT}"; then
+    if ! bazel run //cluster/images/kubemark:push --define REGISTRY="${FULL_REGISTRY}"; then
```
`if ! bazel run //cluster/images/kubemark:push --define REGISTRY="${REGISTRY}" --define REPO_PATH_BASE ="${REPO_PATH_BASE}"; then`

thockin replied:
I prefer it to be simpler - one variable.

I'll leave refactoring of this to the team that owns it ( @shyamjvs and @wojtek-t can overrule if needed)

thockin commented:
squashed.

cblecker commented:
/lgtm

shyamjvs commented:
Oops.. seems like I failed to stop it on time. IIUC this would create problems with our kubemark setup. See point 3 on https://github.com/kubernetes/kubernetes/issues/59567.

shyamjvs commented:
Sorry, I think I misunderstood the change. You seem to define a FULL_REGISTRY that composes REGISTRY and PROJECT, and those value are taken from [here](https://github.com/kubernetes/kubernetes/blob/3586986416b174d390376311c909d665416bf5c0/test/kubemark/cloud-provider-config.sh#L18-L19) which luckily also continues to take the same previous values [here](https://github.com/kubernetes/kubernetes/blob/3586986416b174d390376311c909d665416bf5c0/test/kubemark/start-kubemark.sh#L398-L399). 

I'll however send a PR to unify these values into one single source of truth to avoid bugs in future.

dekkagaijin commented:
@shyamjvs would be good to split on the first '/' of the new single source of truth, when passing values to `bazel run //cluster/images/kubemark:push` in `test/kubemark/start-kubemark.sh`:
https://github.com/kubernetes/kubernetes/blob/master/cluster/images/kubemark/BUILD#L12
https://github.com/bazelbuild/rules_docker#container_push
